File: fsharp/robot-simulator/RobotSimulator.fs
module RobotSimulator

open System.Numerics

let Direction =
    {| North = Complex.ImaginaryOne
       East = Complex.One
       South = -Complex.ImaginaryOne
       West = -Complex.One |}

type Direction = Complex

type Position = Complex

type Robot =
    { direction: Direction
      position: Position }

let create direction (real, imag) =
    { direction = direction
      position = Complex((float)real, (float)imag) }

let move instructions robot =
    Seq.fold (fun r i ->
        match i with
        | 'R' -> { r with direction = r.direction * -Complex.ImaginaryOne }
        | 'L' -> { r with direction = r.direction * Complex.ImaginaryOne }
        | 'A' -> { r with position = r.position + r.direction }
        | _ -> failwith "Unrecognized instruction") robot instructions

File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator
  (:use [clojure.core]))

(def directions [:north :east :south :west])

(def bearing->delta {:north {:x 0 :y 1}
                     :east {:x 1 :y 0}
                     :south {:x 0 :y -1}
                     :west {:x -1 :y 0}})

(defn turn-right [bearing]
  (bearing (zipmap directions (take 4 (drop 1 (cycle directions))))))

(defn turn-left [bearing]
  (bearing (zipmap directions (take 4 (drop 3 (cycle directions))))))

(defn advance [coordinates bearing]
  (merge-with + (bearing->delta bearing) coordinates))

(defn robot [coordinates bearing]
  {:coordinates coordinates :bearing bearing})

(defn robot-operate [robot instruction]
  (case instruction
    \R (update robot :bearing turn-right)
    \L (update robot :bearing turn-left)
    \A (update robot :coordinates advance (:bearing robot))))

(defn simulate [instructions robot]
  (loop [robot robot
         instructions (seq instructions)]
    (if (empty? instructions)
      robot
      (recur (robot-operate robot (first instructions))
             (rest instructions)))))